packages/server-wallet/src/chain-service/event-tracker.ts

```diff
@@ -0,0 +1,105 @@
+  ChallengeRegisteredArg,
+  ChainEventSubscriberInterface,
+} from './types';
+
+/**
+ * This class is a wrapper of ChainEventSubscriberInterface.
+ * Purpose - 2 functions can trigger AssetHolder events to be played to ChainEventSubscriberInterface
+ *    1. chainservice.getInitialHoldings
+ *    2. chainservice.addAssetHolderObservable
+ * Some events can be played twice by both functions, order not predictable,
+ * this class ensures each event is played once and in the right order
+ */
+export class EventTracker {
+  private logger: Logger;
+  private assetHolderMap: Map<Address, {blockNumber: number; logIndex: number}>;
+  private managedSubscriber: ChainEventSubscriberInterface;
+
+  private isNewEvent(assetHolderAddress: Address, blockNumber: number, logIndex: number): boolean {
+    this.logger.debug(
+      `return.isNewEvent: ${assetHolderAddress}, ${blockNumber}, ${logIndex}, ${JSON.stringify(
+        this.assetHolderMap.get(assetHolderAddress) || {}
+      )}`
+    );
+    if (!this.assetHolderMap.has(assetHolderAddress)) {
+      this.assetHolderMap.set(assetHolderAddress, {blockNumber: blockNumber, logIndex: logIndex});
+      return true;
+    } else if (blockNumber > (this.assetHolderMap.get(assetHolderAddress)?.blockNumber || 0)) {
+      this.assetHolderMap.set(assetHolderAddress, {blockNumber: blockNumber, logIndex: logIndex});
+      return true;
+    } else if (blockNumber == (this.assetHolderMap.get(assetHolderAddress)?.blockNumber || 0)) {
+      if (logIndex > (this.assetHolderMap.get(assetHolderAddress)?.logIndex || 0)) {
+        this.assetHolderMap.set(assetHolderAddress, {blockNumber: blockNumber, logIndex: logIndex});
+        return true;
+      } else {
+        return false;
+      }
+    } else {
+      return false;
+    }
+  }
+
+  constructor(managedSubscriber: ChainEventSubscriberInterface, logger: Logger) {
+    this.logger = logger;
+    this.managedSubscriber = managedSubscriber;
+    this.logger = logger;
+    this.assetHolderMap = new Map<string, {blockNumber: number; logIndex: number}>();
+  }
+
+  // Pass event to managed subscriber only if new
+  holdingUpdated(
+    arg: HoldingUpdatedArg,
+    blockNumber: number,
+    logIndex: number = Number.MAX_SAFE_INTEGER // if no logIndex passed in then this is from getInitialHoldings
+  ): void {
+    if (this.isNewEvent(arg.assetHolderAddress, blockNumber, logIndex)) {
+      this.managedSubscriber.holdingUpdated(arg);
+      this.logger.debug(
+        `EventTracker.holdingUpdated: Picked up event (${blockNumber}, ${logIndex}): ` +
+          JSON.stringify(arg)
+      );
+    } else {
+      this.logger.debug(
+        `EventTracker.holdingUpdated: Discarded event (${blockNumber}, ${logIndex}): ` +
+          JSON.stringify(arg)
+      );
+    }
+  }
+
+  // Pass event to managed subscriber only if new
+  assetOutcomeUpdated(
+    arg: AssetOutcomeUpdatedArg,
+    blockNumber: number,
+    logIndex: number = Number.MAX_SAFE_INTEGER
```

Comment:
I am curious on why this default value is needed.

Reply:
getInitialHoldings will read a block instead of an event. In that case, it doesn't have logIndex. However, it should be considered the last event in this block, because the balance it red is final for that block.

Reply:
That makes sense. Would you mind leaving a comment explaining this? Also, do we need this default value for `assetOutcomeUpdated` function? Or should `logIndex` be a required parameter for this function?